### PR TITLE
Implementing direct getwithdrawalshistory method

### PR DESCRIPTION
**Pull request by vazha**
# PR Description

The purpose of this PR is to have ability to get withdrawals history if I don't use a DB option. Because all of presented methods used DB: WithdrawalEventByID, WithdrawalEventsByExchange, WithdrawalEventsByDate.
I certainly thought to create a new API endpoint specifically for this, but perhaps four methods for withdrawals history is already overkill.
Anyway, I wanted to discuss this with you.
Аt the moment I using an additional parameter - currency in the RPC WithdrawalEventsByExchange method to make direct requests to exchanges without using a database.
Аt the moment I have implemented this method for Kraken and Binance exchanges only.

Fixes # (issue)
Some typos in Kraken methods

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## How has this been tested

- [x] go test ./... -race

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code

### CODE REVIEW

**Review by MadCozBadd — COMMENTED** [review]
Nice work m8888888. I cant really comment on the design aspect as to whether this function/method should be added or not but heres a few changes from me assuming others decide to approve the method.

Note - you need to add this new function to ibotexchange in interfaces.go

**Review by shazbert — CHANGES_REQUESTED** [review]
Thanks for the PR! Need to add a method to FakePassingExchange in fake_exchange_test.go, and some small nits below.

File: exchanges/kraken/kraken_wrapper.go [hunk 1]
```diff
@@ -516,6 +516,24 @@ func (k *Kraken) GetFundingHistory() ([]exchange.FundHistory, error) {
 	return nil, common.ErrFunctionNotSupported
 }
 
+// GetWithdrawalsHistory returns previous withdrawals data
+func (k *Kraken) GetWithdrawalsHistory(currency currency.Code) (resp []exchange.WithdrawalHistory, err error) {
+	withdrawals, err := k.WithdrawStatus(currency, "")
+	for _, w := range withdrawals {
```
[the remark this answers is not in the document]
Comment (vazha) [reply exchanges/kraken/kraken_wrapper.go:522]: Great! Thanks for all your comments!

File: exchanges/binance/binance.go [hunk 2]
```diff
@@ -746,6 +746,24 @@ func (b *Binance) WithdrawCrypto(asset, address, addressTag, name, amount string
 	return resp.ID, nil
 }
 
+// WithdrawStatus gets the status of recent withdrawals
+func (b *Binance) WithdrawStatus(c currency.Code) ([]WithdrawStatusResponse, error) {
```
Comment (gloriousCode) [inline exchanges/binance/binance.go:750]: While in the context of the rpc function we don't use it yet, I would appreciate if you filled out the functionality of this implementation to match Binance's wAPI https://github.com/binance-exchange/binance-official-api-docs/blob/master/wapi-api.md#withdraw-history-user_data
ie add `startTime`, `endTime`, `status`

File: exchanges/binance/binance_wrapper.go [hunk 3]
```diff
@@ -498,6 +498,29 @@ func (b *Binance) GetFundingHistory() ([]exchange.FundHistory, error) {
...
 }
 
+// GetWithdrawalsHistory returns previous withdrawals data
+func (b *Binance) GetWithdrawalsHistory(c currency.Code) (resp []exchange.WithdrawalHistory, err error) {
+	w, err := b.WithdrawStatus(c)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, b := range w {
```
Comment (gloriousCode) [inline exchanges/binance/binance_wrapper.go:508]: Would you be able to update this to be an index? `for i := range w {`

File: exchanges/coinbasepro/coinbasepro_wrapper.go [hunk 4]
```diff
@@ -439,7 +439,7 @@ func (c *CoinbasePro) GetFundingHistory() ([]exchange.FundHistory, error) {
 }
 
 // GetWithdrawalsHistory returns previous withdrawals data
-func (c *CoinbasePro) GetWithdrawalsHistory(currency currency.Code) (resp []exchange.WithdrawalHistory, err error) {
+func (c *CoinbasePro) GetWithdrawalsHistory(c currency.Code) (resp []exchange.WithdrawalHistory, err error) {
```
Comment (shazbert) [inline exchanges/coinbasepro/coinbasepro_wrapper.go:442]: Duplicate variable declaration.

File: exchanges/coinbene/coinbene_wrapper.go [hunk 5]
```diff
@@ -508,7 +508,7 @@ func (c *Coinbene) GetFundingHistory() ([]exchange.FundHistory, error) {
 }
 
 // GetWithdrawalsHistory returns previous withdrawals data
-func (c *Coinbene) GetWithdrawalsHistory(currency currency.Code) (resp []exchange.WithdrawalHistory, err error) {
+func (c *Coinbene) GetWithdrawalsHistory(c currency.Code) (resp []exchange.WithdrawalHistory, err error) {
```
Comment (shazbert) [inline exchanges/coinbene/coinbene_wrapper.go:511]: Duplicate variable declaration.

File: exchanges/coinut/coinut_wrapper.go [hunk 6]
```diff
@@ -501,7 +501,7 @@ func (c *COINUT) GetFundingHistory() ([]exchange.FundHistory, error) {
 }
 
 // GetWithdrawalsHistory returns previous withdrawals data
-func (c *COINUT) GetWithdrawalsHistory(currency currency.Code) (resp []exchange.WithdrawalHistory, err error) {
+func (c *COINUT) GetWithdrawalsHistory(c currency.Code) (resp []exchange.WithdrawalHistory, err error) {
```
Comment (shazbert) [inline exchanges/coinut/coinut_wrapper.go:504]: Duplicate variable declaration.

File: exchanges/kraken/kraken_wrapper.go [hunk 7]
```diff
@@ -516,6 +516,24 @@ func (k *Kraken) GetFundingHistory() ([]exchange.FundHistory, error) {
...
+	withdrawals, err := k.WithdrawStatus(c, "")
+	for i := range withdrawals {
+		resp = append(resp, exchange.WithdrawalHistory{
+			Status:          withdrawals[i].Status,
+			TransferID:      withdrawals[i].Refid,
+			Timestamp:       time.Unix(int64(withdrawals[i].Time), 0),
+			Amount:          withdrawals[i].Amount,
+			Fee:             withdrawals[i].Fee,
+			CryptoToAddress: withdrawals[i].Info,
+			CryptoTxID:      withdrawals[i].TxID,
```
Comment (gloriousCode) [inline exchanges/kraken/kraken_wrapper.go:530]: Since you add the currency for Binance;s implementation, you may as well add it here too

**Review by shazbert — CHANGES_REQUESTED** [review]
Thanks for making those changes, found some issues on the next pass, and then once: https://github.com/thrasher-corp/gocryptotrader/pull/600#discussion_r530697356 is complete then this looks good to go.

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thanks for this! We shouldn't need to rely on just the database withdrawal entries, so this is a good addition. I've tested both API endpoints and they work. Just a few comments

File: exchanges/coinut/coinut_wrapper.go (hunk 6 above)
Comment (vazha) [reply exchanges/coinut/coinut_wrapper.go:504]: Thanks!

File: exchanges/kraken/kraken_wrapper.go (hunk 7 above)
Comment (vazha) [reply exchanges/kraken/kraken_wrapper.go:530]: Thanks, done.

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thanks for making those changes! Last request is to add test coverage to the `binance.go` function `WithdrawStatus`. We do have mock testing for Binance (which you can read about [here](https://github.com/thrasher-corp/gocryptotrader/tree/master/exchanges/mock)), so if you have any difficulties, let us know and we can help out!

**Comment by gloriousCode** [comment]
Actually another thing I just noticed, there's a build error: https://ci.appveyor.com/project/thrasher-/gocryptotrader-g5pk0/builds/36545026#L55
If you could address that one too. The fake testing exchange requires a basic implementation of `GetWithdrawalsHistory`

**Review by shazbert — COMMENTED** [review]
Changes look good!

**Review by thrasher- — CHANGES_REQUESTED** [review]
Please rebase/merge master

**Comment by vazha** [comment]
> Thanks for making those changes! Last request is to add test coverage to the `binance.go` function `WithdrawStatus`. We do have mock testing for Binance (which you can read about [here](https://github.com/thrasher-corp/gocryptotrader/tree/master/exchanges/mock)), so if you have any difficulties, let us know and we can help out!

Thanks, I added the test, please check if it's OK

**Review by MadCozBadd — APPROVED** [review]
tACK 🥳

File: engine/rpcserver.go [hunk 8]
```diff
@@ -1303,6 +1303,20 @@ func (s *RPCServer) WithdrawalEventByID(_ context.Context, r *gctrpc.WithdrawalE
 // WithdrawalEventsByExchange returns previous withdrawal request details by exchange
 func (s *RPCServer) WithdrawalEventsByExchange(_ context.Context, r *gctrpc.WithdrawalEventsByExchangeRequest) (*gctrpc.WithdrawalEventsByExchangeResponse, error) {
 	if !s.Config.Database.Enabled {
+		if r.Id == "" {
```
Comment (gloriousCode) [inline engine/rpcserver.go:1306]: I think I'd prefer `gctrpc.WithdrawalEventsByExchangeRequest` to contain a property which selects whether to go via API or database here. Essentially: 
```
.\gctcli.exe withdrawalrequesthistory byexchange --exchange=binance --currency=USDT --source=api
```
Where `source` is either `database` or `api` and that determines which endpoint to select here. This means that I can still retrieve withdrawalhistory if I want if I have the database enabled.

What do you think?

**Review by gloriousCode — COMMENTED** [review]
Hey Vahza, I've been doing some testing and it all works well, thank you for making those changes!

I've got one comment, its not a blocker, but I'm curious of your thoughts

File: engine/rpcserver.go (hunk 8 above)
Comment (vazha) [reply engine/rpcserver.go:1306]: Hello! :) I'm not sure if this won't confuse users, for example if someone sends request "--currency=USDT --source=database" and expects to get only USDT results, but that wouldn't be so now. In that case need to change appropriate functions to filter DB results by currency, right?

Comment (gloriousCode) [reply engine/rpcserver.go:1306]: Fair enough! I didn't notice the lack of currency filter in that function! So nevermind. That's a bit more work to achieve and can always be done later. tACK!

**Review by gloriousCode — APPROVED** [review]

**Review by thrasher- — APPROVED** [review]
tACK, thanks @vazha !
image

**Comment by thrasher-** [comment]
Before merging, just noticed a linter issue: https://ci.appveyor.com/project/thrasher-/gocryptotrader-g5pk0/builds/36595224#L56

**Review by gloriousCode — APPROVED** [review]
ACK latest changes

**Review by shazbert — APPROVED** [review]
Tested approved.

**Review by thrasher- — APPROVED** [review]

**Pull request merged**